### ISSUE TITLE
API-wide authentication

### ISSUE THREAD

**Issue by karthikb351** [aneesh-neelam/VITacademics#12]
We should do a basic auth of getting the regno+dob for critical requests.

Things like token refresh and generation currently only require a regno, which is basically no security whatsoever and vulnerable to both exploits as well as spamming.
### Vulnerable Endpoints

```
- /api/vellore/login/manual?regno=(regno)
- /api/vellore/data/first?regno=(regno)
- /api/vellore/data/refresh?regno=(regno)
- /api/vellore/friends/regenerate?regno=(regno)
```
### Possible Solution

When a user hits the `/api/vellore/data/first?regno=(regno)` endpoint, if the details are correct and the data is parsed, we store his creds in our db so we can use it to verify all subsequent requests.
(I'm aware this is against what I said earlier about storing user data, but I can't think of any other way)

**Comment by aneesh-neelam** [comment]
I had thought of this before. Didn't do it cause I found it too much effort at the time. 

Anyway, I think storing the DoB in the cache along with the cookie would be faster than using MongoDB. [node-cache](https://github.com/ptarjan/node-cache) can store anything: objects, arrays, strings etc. I can store the DoB with the cookie. 

In anycase, I'll get right on it. Consider it a high priority. 

Only the data and friends API calls will be affected. Manual Captcha Login already has submit DoB, not necessary for fetching Captcha. 

New API Endpoints: 
- /api/vellore/data/first?regno=(regno)&dob=(dob)
- /api/vellore/data/refresh?regno=(regno)&dob=(dob)
- /api/vellore/friends/regenerate?regno=(regno)&dob=(dob)
- /api/chennai/data/first?regno=(regno)&dob=(dob)
- /api/chennai/data/refresh?regno=(regno)&dob=(dob)
- /api/chennai/friends/regenerate?regno=(regno)&dob=(dob)

@biocross, when you're writing the new app, please use the new endpoints. Just send dob along with regno.

**Comment by karthikb351** [comment]
Once we've verified that the dob+regno combination is correct (which should happen if `/api/vellore/data/first?regno=(regno)&dob=(dob)` goes through) then just hash and store in cache. We can use the hash to quickly check if the combination is correct then.

**Comment by aneesh-neelam** [comment]
Cache is only temporary, think of it as volatile storage. Valid for only a few minutes, but enough for getting captcha, logging in and getting data from VIT's server as their cookie is anyway not valid for long. 

I was thinking to store the DoB in the cache along with the cookie, valid for the same amount of time, to be used only for authentication. 

And for regenerating token, MongoDB can be used. Regenerating token is an internal thing, so does not depend on how fast VIT's server and connection are. 

And what's the point of hashing when we're storing everything? :-P

**Comment by aneesh-neelam** [comment]
Done! Fixed in ff856b388239d6e299d6e38871f06e7429bb1dbf
Bug fixes in d5618c244c7a0d4220ee3d2afff96056cf804516
